### PR TITLE
TST: fix bundle test discovery

### PR DESCRIPTION
**Pull request by llllllllll**

### CODE REVIEW

File: tests/data/bundles/test_core.py [hunk 1]
```diff
@@ -303,7 +303,7 @@ def _(environ,
                   output_dir):
                 _wrote_to.append(output_dir)
 
-        _wrote_to.clear()
+        _wrote_to[:] = []
```
Comment (richafrank) [inline tests/data/bundles/test_core.py]: What was this change for?  Also, is it intentional that the default value for `_wrote_to` is mutable?

**Comment by richafrank** [comment]
Looks like travis is failing with a precision issue, and appveyor is failing with #1228.

File: tests/data/bundles/test_core.py (hunk 1 above)
Comment (llllllllll) [inline tests/data/bundles/test_core.py]: clear was added to list in python 3, it is supposed to be mutable.

**Comment by linc09** [comment]
I also find this problem when ingest yahoo at windows 7, the error msg is 
"
sqlite3.OperationalError: unable to open database file
"
then I find when run
`            if overwrite and exists(conn_or_path):
                try:
                    remove(conn_or_path)
                except OSError as e:
                    if e.errno != ENOENT:
                        raise`
the 'conn_or_path' file still exist after remove.

**Comment by richafrank** [comment]
(This is in service of fixing https://github.com/quantopian/zipline/issues/1228.)

File: zipline/utils/cache.py [hunk 2]
```diff
@@ -298,9 +304,25 @@ def __enter__(self):
         return self
 
     def __exit__(self, *exc_info):
+        self._tmpfile.__exit__(*exc_info)
         if exc_info[0] is None:
             self._commit()
-        self._tmpfile.__exit__(*exc_info)
+
+    @property
```
Comment (llllllllll) [inline zipline/utils/cache.py]: why are we using getters and setters here instead of just naming the attribute `tmpfile`?

File: zipline/data/bundles/core.py [hunk 3]
```diff
@@ -353,14 +353,19 @@ def ingest(name,
                         asset_db_path(name, timestr, environ=environ),
                     )).path,
                 )
+                wf = stack.enter_context(working_file(
+                    adjustment_db_path(name, timestr, environ=environ),
+                ))
+                # close the file to delete it on windows
```
Comment (llllllllll) [inline zipline/data/bundles/core.py]: I don't think this comment is still correct.

File: zipline/__init__.py [hunk 4]
```diff
@@ -12,6 +12,24 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import os
+
+
+if os.name == 'nt':
```
Comment (llllllllll) [inline zipline/__init__.py]: does this still matter?

File: zipline/data/us_equity_pricing.py [hunk 5]
```diff
@@ -1021,6 +1020,12 @@ def write_dividend_data(self, dividends, stock_dividends=None):
         dividend_ratios = self.calc_dividend_ratios(dividends)
         self.write_frame('dividends', dividend_ratios)
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
```
Comment (llllllllll) [inline zipline/data/us_equity_pricing.py]: where do we enter this?

File: zipline/utils/cache.py [hunk 6]
```diff
@@ -288,7 +289,12 @@ def path(self):
     def _commit(self):
         """Sync the temporary file to the final path.
         """
-        copyfile(self.name, self._final_path)
+        self._tmpfile.close()
+        try:
+            move(self._name, self._final_path)
+        except OSError as e:
+            if e != errno.ENOENT:
```
Comment (llllllllll) [inline zipline/utils/cache.py]: shouldn't this fail if the file doesn't exist?

File: zipline/utils/cache.py (hunk 2 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: No reason in particular; if the tmpfile isn't necessarily a closeable file, then I wanted to be able to handle that in the future. 

In this case, the file is actually an SQLite3 Connection object. I figured to add these just in case there was some unforeseeable object type that the working file needed to use.

Comment (llllllllll) [inline zipline/utils/cache.py]: if we need to handle that in the future then we can swap to a property. I think that for now it is more clear to just expose the attribute.

File: zipline/utils/cache.py (hunk 6 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: Yeah, didn't even realize that was still in there

File: zipline/utils/cache.py (hunk 2 above)
Comment (llllllllll) [inline zipline/utils/cache.py]: also, do we ever need to access the tmpfile directly, we forward all attrs to the tmpfile through our `__getattribute__`, so `working_file.close` is the same as `workingfile.tmpfile.close`

File: zipline/data/us_equity_pricing.py (hunk 5 above)
Comment (LotannaEzenwa) [inline zipline/data/us_equity_pricing.py]: we don't; I added it because there's an exit.

File: zipline/__init__.py (hunk 4 above)
Comment (LotannaEzenwa) [inline zipline/__init__.py]: I think so; if this is making all the tmp files go to this directory, it is necessary. The tmp files have auto-delete turned off, so if they're ending up here, it's necessary for cleanup

**Comment by LotannaEzenwa** [comment]
@llllllllll  These should all pass, can you take a look at the refactored code?

File: tests/data/bundles/test_yahoo.py [hunk 7]
```diff
@@ -1,11 +1,13 @@
+from __future__ import division
```
Comment (llllllllll) [inline tests/data/bundles/test_yahoo.py]: are we doing any division in here?

File: zipline/data/bundles/core.py [hunk 8]
```diff
@@ -38,39 +37,59 @@
 
 def asset_db_path(bundle_name, timestr, environ=None):
     return pth.data_path(
-        [bundle_name, timestr, 'assets-%d.sqlite' % ASSET_DB_VERSION],
-        environ=environ,
+        list(asset_db_relative(bundle_name, timestr, environ)),
```
Comment (llllllllll) [inline zipline/data/bundles/core.py]: do we need to call list on these tuples? Can the input be any iterable?

File: zipline/utils/cache.py [hunk 9]
```diff
@@ -350,7 +373,7 @@ def getpath(self, *path_parts):
     def _commit(self):
         """Sync the temporary directory to the final path.
         """
-        copytree(self.path, self._final_path)
+        dir_util.copy_tree(self.path, self._final_path, verbose=1)
```
Comment (llllllllll) [inline zipline/utils/cache.py]: what is this change?

File: zipline/utils/cache.py [hunk 10]
```diff
@@ -334,7 +353,11 @@ def mkdir(self, *path_parts):
             The parts of the path after the working directory.
         """
         path = self.getpath(*path_parts)
-        os.mkdir(path)
+        try:
```
Comment (llllllllll) [inline zipline/utils/cache.py]: I would call this function `ensure_dir` because we have changed the semantics away from `mkdir`. We could have both, where `ensure_dir` is implemented in terms of `mkdir`. What do you think?

File: zipline/utils/cache.py (hunk 2 above)
Comment (llllllllll) [inline zipline/utils/cache.py]: we can now remove these, right?

File: zipline/testing/fixtures.py [hunk 11]
```diff
@@ -651,6 +655,10 @@ def make_equity_daily_bar_data(cls):
             )
 
     @classmethod
+    def make_bcolz_daily_bar_rootdir_path(cls):
```
Comment (llllllllll) [inline zipline/testing/fixtures.py]: what is this change?

File: zipline/data/bundles/core.py [hunk 12]
```diff
@@ -321,53 +340,56 @@ def ingest(name,
...
-                    )).path,
-                )
-                adjustment_db_writer = SQLiteAdjustmentWriter(
-                    stack.enter_context(working_file(
-                        adjustment_db_path(name, timestr, environ=environ),
-                    )).path,
-                    BcolzDailyBarReader(daily_bars_path),
-                    bundle.calendar,
-                    overwrite=True,
+                asset_db_writer = stack.enter_context(
```
Comment (llllllllll) [inline zipline/data/bundles/core.py]: thanks for entering these now to ensure the connection is closed!

File: zipline/data/bundles/core.py [hunk 13]
```diff
@@ -321,53 +340,56 @@ def ingest(name,
...
+                    AssetDBWriter(
+                        wd.getpath(*asset_db_relative(
+                            name, timestr, environ=environ)
```
Comment (llllllllll) [inline zipline/data/bundles/core.py]: can we align the closing paren with the indentation level of the line of the opening paren. For example:

```python
AssetDBWriter(
    wd.getpath(*asset_db_relative(
        name, timestr, environ=environ,
    ),
    ...
)
```

File: tests/data/bundles/test_yahoo.py (hunk 7 above)
Comment (LotannaEzenwa) [inline tests/data/bundles/test_yahoo.py]: Already merged with origin/master, so a few of these updates are from the recent PR's

File: zipline/data/bundles/core.py (hunk 8 above)
Comment (LotannaEzenwa) [inline zipline/data/bundles/core.py]: Not necessarily, I made these a list so the underlying logic of paths.py wouldn't be changed

File: zipline/utils/cache.py (hunk 9 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: copytree() fails if the target directory already exists on py2.7 but not 3.4, dir_util.copy_tree is okay with merging on both versions of python

File: zipline/utils/cache.py (hunk 10 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: If `ensure_dir` makes subdirectories when the parent doesn't exist, then that change will work. The reason for using `makedirs` as opposed to `mkdir` is that `mkdir` fails when the parents don't exist.

File: zipline/utils/cache.py (hunk 9 above)
Comment (llllllllll) [inline zipline/utils/cache.py]: cool, why is it verbose though?

File: zipline/testing/fixtures.py (hunk 11 above)
Comment (LotannaEzenwa) [inline zipline/testing/fixtures.py]: From the merge with o/m

File: zipline/utils/cache.py (hunk 9 above)
Comment (llllllllll) [inline zipline/utils/cache.py]: also, the class docstring needs to be updated to reflect this change

File: zipline/testing/fixtures.py (hunk 11 above)
Comment (llllllllll) [inline zipline/testing/fixtures.py]: this shouldn't appear as a diff in your pr though, did you rebase?

File: zipline/data/us_equity_pricing.py (hunk 5 above)
Comment (LotannaEzenwa) [inline zipline/data/us_equity_pricing.py]: Now we enter this, as above

File: zipline/utils/cache.py (hunk 10 above)
Comment (llllllllll) [inline zipline/utils/cache.py]: Sure, I was more concerned with the fact that this will not fail if the directory already exists. Also, we have `paths.ensure_directory` which should do this.

File: zipline/assets/asset_writer.py [hunk 14]
```diff
@@ -563,3 +560,9 @@ def _load_data(self, equities, futures, exchanges, root_symbols):
             exchanges=exchanges_output,
             root_symbols=root_symbols_output,
         )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        pass
```
Comment (richafrank) [inline zipline/assets/asset_writer.py]: Should this be closing a connection here?

File: zipline/utils/cache.py (hunk 9 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: It was verbose for testing, but yeah it should be suppressed now.

File: zipline/utils/cache.py (hunk 10 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: Ah, I see. There's actually no possible way for the directory to exist because it's within a temp directory...I think. I could be wrong. 

But yeah you're right about it not failing, that was for catching the errors in testing.

File: zipline/testing/fixtures.py (hunk 11 above)
Comment (LotannaEzenwa) [inline zipline/testing/fixtures.py]: I rebased ~35 of my commits before merging

File: zipline/data/bundles/core.py (hunk 13 above)
Comment (LotannaEzenwa) [inline zipline/data/bundles/core.py]: With 2 paren, it would look like this?
```python
AssetDBWriter(
    wd.getpath(*asset_db_relative(
        name, timestr, environ=environ,
    )),
)
```

File: zipline/utils/cache.py (hunk 2 above)
Comment (LotannaEzenwa) [inline zipline/utils/cache.py]: Yeah, working file isn't used in ingest anymore

File: zipline/assets/asset_writer.py (hunk 14 above)
Comment (LotannaEzenwa) [inline zipline/assets/asset_writer.py]: The only thing that the cm is capable of closing is the SqlAlchemy engine, which has a dispose method. It closes all the checked-in connections. All of the checked out connections will be left open, however. 

I'm not exactly sure if this method should try to close all connections, or just those that are not in use

**Comment by LotannaEzenwa** [comment]
@llllllllll , @richafrank fixed up some last styles/comments.

File: zipline/assets/asset_writer.py [hunk 15]
```diff
@@ -563,3 +560,9 @@ def _load_data(self, equities, futures, exchanges, root_symbols):
             exchanges=exchanges_output,
             root_symbols=root_symbols_output,
         )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        pass  # Shouldn't dispose of the engine explicitly
```
Comment (richafrank) [inline zipline/assets/asset_writer.py]: Do you think there's still a benefit to this becoming a context manager if it hasn't nothing to clean up?

File: tests/data/bundles/test_yahoo.py (hunk 7 above)
Comment (richafrank) [inline tests/data/bundles/test_yahoo.py]: Just a note: before merging the PR, we should clean this up.  Ideally. we shouldn't be seeing unrelated changes from other PRs.

File: zipline/assets/asset_writer.py (hunk 15 above)
Comment (LotannaEzenwa) [inline zipline/assets/asset_writer.py]: Functionally, no. 

But for the code as a whole, I think it makes sense for all of the writer objects to be context managers as they control system resources. 
This asset writer is a special case because it's querying system doesn't need to be cleaned up explicitly.
If someone were to change the asset writer's underlying system for performance or to remove redundancies, it would be easy to add the exit method if needed.

Comment (richafrank) [inline zipline/assets/asset_writer.py]: Ok, two things then:

1. If we want the asset writer to be a ctx manager, then we should update all consumers to dispose of it when they're done.  Right now, I think we've only updated the zipline data bundle.  However, given that it doesn't actually manage any resources right now that need explicit disposal, my tendency is to undo this and re-do it with all the call sites updated if/when there are resources to clean up.  @llllllllll What do you think?
1. We should also update this comment to explain why we shouldn't dispose of the engine.  For example, will it cause problems if we do, or is it just unnecessary?

Comment (llllllllll) [inline zipline/assets/asset_writer.py]: Can you explain more why we don't need to dispose of the engine? I afree that we should either undo this change, or make it actually dispose of the engine and update the callers to use the ctx manager aspect of the object.

File: zipline/data/us_equity_pricing.py [hunk 16]
```diff
@@ -210,8 +210,11 @@ class BcolzDailyBarWriter(object):
     def __init__(self, filename, calendar, start_session, end_session):
         self._filename = filename
 
-        assert calendar.is_session(start_session), "Start session is invalid!"
-        assert calendar.is_session(end_session), "End session is invalid!"
+        if start_session != end_session:
+            assert calendar.is_session(start_session),\
```
Comment (llllllllll) [inline zipline/data/us_equity_pricing.py]: Can we make these a `ValueError` and show the session in the message. In general, assertions should be used to validate invariants in code that do not depend on user inputs. An assertion error should indicate that there is a bug in the program, but not that a function was called with invalid arguments.

**Comment by llllllllll** [comment]
This looks great, thanks for getting this fixed! Other than the question about the context manager, we just need to clean up the commit messages and then this should be good to merge.

File: zipline/data/us_equity_pricing.py (hunk 16 above)
Comment (LotannaEzenwa) [inline zipline/data/us_equity_pricing.py]: Okay, fixing it

File: zipline/assets/asset_writer.py (hunk 15 above)
Comment (LotannaEzenwa) [inline zipline/assets/asset_writer.py]: Well I took another look at the overall docs for the sqlalchemy engine, and it said that the engine shouldn't be disposed until the end of the program because this prevents any future queries to the database. 
As of right now, the readers don't try to explicitly create an engine, so closing the engine would cause the readers to fail.

 I agree that the change wouldn't be necessary, as of right now

File: tests/data/bundles/test_yahoo.py (hunk 7 above)
Comment (richafrank) [inline tests/data/bundles/test_yahoo.py]: @llllllllll thinks he added this to the PR intentionally to make this module compatible with both python 2 and 3.

**Comment by LotannaEzenwa** [comment]
Merging
